extensions/basic_auth/plugin.h

```diff
@@ -0,0 +1,92 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "extensions/common/context.h"
+#include "google/protobuf/util/json_util.h"
+
+#ifndef NULL_PLUGIN
+
+#include <assert.h>
+#define ASSERT(_X) assert(_X)
+
+#include "proxy_wasm_intrinsics.h"
+
+static const std::string EMPTY_STRING;
+
+#else
+
+#include "include/proxy-wasm/null_plugin.h"
+
+namespace proxy_wasm {
+namespace null_plugin {
+namespace BasicAuth {
+namespace Plugin {
+
+#endif
+
+using google::protobuf::util::JsonParseOptions;
+using google::protobuf::util::Status;
+
+// PluginRootContext is the root context for all streams processed by the
+// thread. It has the same lifetime as the worker thread and acts as target for
+// interactions that outlives individual stream, e.g. timer, async calls.
+class PluginRootContext : public RootContext {
+ public:
+  PluginRootContext(uint32_t id, std::string_view root_id)
+      : RootContext(id, root_id) {}
+  ~PluginRootContext() {}
+  bool onConfigure(size_t) override;
+
+  enum MATCH_TYPE { Prefix, Exact, Suffix };
+  struct BasicAuthConfigRule {
+    std::string request_path;
+    MATCH_TYPE pattern;
+    std::unordered_set<std::string> encoded_credentials;
+  };
+  const std::unordered_map<std::string,
+                           std::vector<PluginRootContext::BasicAuthConfigRule>>
+  basicAuthConfigurationValue() {
+    return basic_auth_configuration_;
+  };
+
+ private:
+  bool configure(size_t);
+  std::unordered_map<std::string,
+                     std::vector<PluginRootContext::BasicAuthConfigRule>>
+      basic_auth_configuration_;
+};
+
+// Per-stream context.
+class PluginContext : public Context {
+ public:
+  explicit PluginContext(uint32_t id, RootContext* root) : Context(id, root) {}
+  FilterHeadersStatus onRequestHeaders(uint32_t, bool) override;
+
+ private:
+  FilterHeadersStatus credentialsCheck(
```

Comment:
Instead of implementing this in PluginContext, you can just implement it in RootContext, so that you can avoid copying configuration from it. When PluginContext needs to do credential check, it can just call `rootContext()->checkCredential(...)`.

Reply:
oh i see! done.

Reply:
Hmm I might not be clearly here. I think the whole logic inside `onRequestHeaders` could just be moved into a method defined in RootContext (you can call it check or something). Then you don't need to exchange data structure between root context and plugin context. Check logic could just contained inside root context.